arirs/src/client.rs

```diff
@@ -2,208 +2,107 @@ use std::time::Duration;
-                                    event!(Level::INFO, "Received WebSocket Pong");
-                                },
-                                tungstenite::Message::Close(frame) => {
-                                    event!(Level::INFO, "WebSocket closed: {:?}", frame);
-                                    break;
-                                },
-                                _ => {
-                                    event!(Level::INFO, "Unknown WebSocket message");
+    async fn connect_ws(ws_url: Url) -> Result<UnboundedReceiver<Event>> {
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+
+        let (ws_stream, _) = connect_async(&ws_url.to_string()).await?;
+
+        let _join_task: JoinHandle<Result<()>> = tokio::task::spawn(async move {
+            let (mut ws_sender, mut ws_receiver) = ws_stream.split();
+            let mut interval = interval(Duration::from_millis(5000));
+
+            loop {
+                tokio::select! {
+                    message = ws_receiver.next() => {
+                        match message {
+                            Some(message) => {
+                                let message = message?;
+                                match message {
+                                    tungstenite::Message::Text(_) => {
+                                        if let Err(e) = tx.send(serde_json::from_slice(message.into_data().as_slice()).map_err(|err| AriError::Unknown(err.to_string()))?) {
+                                            event!(Level::ERROR, "Error sending event: {}", e);
+                                        }
+                                    }
+                                    tungstenite::Message::Ping(data) => {
+                                        event!(Level::TRACE, "Received WebSocket Ping, sending Pong");
+                                        ws_sender.send(tungstenite::Message::Pong(data)).await?;
+                                    }
+                                    tungstenite::Message::Pong(_) => {
+                                        event!(Level::TRACE, "Received WebSocket Pong");
+                                    },
+                                    tungstenite::Message::Close(frame) => {
+                                        event!(Level::INFO, "WebSocket closed: {:?}", frame);
+                                        break;
+                                    },
+                                    _ => {
+                                        event!(Level::INFO, "Unknown WebSocket message");
+                                    }
                                 }
                             }
-                        }
-                        None => {
-                            event!(Level::INFO, "WebSocket closed");
-                            break;
+                            None => {
+                                tracing::error!("WebSocket closed");
```

Comment:
question: is a websocket closing an error?

Reply:
Probably, since we're the ones driving it. I would much rather return an error here. I'll be going over logging soon.